modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm

```diff
@@ -0,0 +1,206 @@
+  my $dba  = $self->get_DBAdaptor($type);
+  my $registry = 'Bio::EnsEMBL::Registry';
+  if(! $dba) {
+    $self->info("Cannot find adaptor for type %s", $type);
+    next;
+  }
+  my $attribute_adaptor = $dba->get_adaptor('Attribute');
+  if( $seq_type eq 'toplevel'){
+    $genome_ad = $registry->get_adaptor( $species, $type, "Slice" );
+    $attrib_table = "seq_region";
+    $genome_feature_id = $self->param('seq_region_id');
+    $genome_slice = $genome_ad->fetch_by_seq_region_id($self->param('seq_region_id'));
+  }
+  elsif($seq_type eq 'cdna' || $seq_type eq 'cds' || $seq_type eq 'pep'){
+    $genome_ad = $registry->get_adaptor($species, $type, "Transcript");
+    $genome_feature = $genome_ad->fetch_by_dbID($self->param('transcript_dbid'));
+    $genome_slice = $genome_feature->seq; #create a seq object 
+    $attrib_table = "transcript";
+    $genome_feature_id = $self->param('transcript_dbid');
+
+    if( $seq_type eq 'cds' ){
+      $genome_slice->seq($genome_feature->translateable_seq);
+    }
+    if( $seq_type eq 'pep' ){
+      $genome_slice = $genome_feature->translate;
+      $attrib_table = "translation";
+      $genome_feature_id = $self->param('translation_dbid');
+    }	    
+  }else{
+    die "Unknown $seq_type : valid seq_types are toplevel, cdna, cds, pep";	  
+  }
+
+  #generate hash and update the attrb table 
+  foreach my $hash_method (@$hash_types){
+    my $at_code = $hash_method."_".$seq_type;
+    my $sequence_hash = $self->generate_sequence_hash($genome_slice, $hash_method);
+    $self->update_attrib_table($attribute_adaptor, $at_code, $sequence_hash, $genome_feature_id , $attrib_table);
+  }  
+}
+
+sub generate_sequence_hash {
+  my ($self, $slice, $hash_method) = @_;
+  my $sequence_hash;
+  if( $hash_method =~ /sha\-?(\d+)/){
+    my $digest_size = $self->param('digest_size') || 24;
+    my $algo = int($1);    
+    $sequence_hash = $self->sha512t24u($slice, $algo, $digest_size);	  
+  }elsif($hash_method eq "md5"){
+     $sequence_hash = $self->md5($slice);	  
+  }	  
+  return $sequence_hash; 
+}
+
+sub sha512t24u {
+  my ($self, $slice, $algo, $digest_size) = @_;
+
+  if(($digest_size % 3) != 0) {
+    die "Digest size must be a multiple of 3 to avoid padded digests";
+  }
+  my $sha = Digest::SHA->new(int($algo)); #even though we use sha512 algorithm we truncate the generated hash based on given digest size  
+  $sha = $self->sequence_stream_digest($slice, $sha, $self->param('chunk_size')); 
+  my $digest = $sha->digest;
+  my $base64 = encode_base64url($digest);
+  my $substr_offset = int($digest_size/3)*4;
+  return substr($base64, 0, $substr_offset);
+}
+
+sub md5{
+  my ($self, $slice) = @_;
+  my $md5 = Digest::MD5->new;
+  $md5 = $self->sequence_stream_digest($slice, $md5, $self->param('chunk_size'));
+  my $digest = $md5->digest;
+  my $base64 = encode_base64url($digest);
+  return $base64 
+}
+
+sub sequence_stream_digest {
```

Comment:
Just double checking that this is working correctly i.e. we can push say the MD5 checksums into ENA's refget and find our sequences?